src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java

```diff
@@ -0,0 +1,251 @@
+                    // Also try to see if these are exactly the contents of a code block
+                    else if ((openBraceOffset == -1) && (closeBraceOffset == -1) && (startOffset > 0) && (endOffset < documentLength)) {
+                        startChar = documentChars.charAt(startOffset - 1);
+                        endChar = documentChars.charAt(endOffset);
+
+                        if (LSPCodeBlockUtils.isCodeBlockStartChar(file, startChar)) {
+                            Character pairedCloseBraceChar = LSPCodeBlockUtils.getCodeBlockEndChar(file, startChar);
+                            if ((pairedCloseBraceChar != null) && (pairedCloseBraceChar == endChar)) {
+                                containingTextRanges.add(textRange);
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Return the closest (i.e., smallest) containing text range
+            if (!ContainerUtil.isEmpty(containingTextRanges)) {
+                containingTextRanges.sort(Comparator.comparingInt(TextRange::getLength));
+                return ContainerUtil.getFirstItem(containingTextRanges);
+            }
+        }
+
+        return null;
+    }
+
+    @Nullable
+    private static TextRange getUsingFoldingRanges(@NotNull PsiFile file,
+                                                   @NotNull Document document,
+                                                   int offset,
+                                                   @Nullable Character openBraceChar,
+                                                   int openBraceOffset,
+                                                   @Nullable Character closeBraceChar,
+                                                   int closeBraceOffset) {
+        List<FoldingRange> foldingRanges = LSPFoldingRangeBuilder.getFoldingRanges(file);
+        if (!ContainerUtil.isEmpty(foldingRanges)) {
+            CharSequence documentChars = document.getCharsSequence();
+            int documentLength = documentChars.length();
+
+            List<TextRange> containingTextRanges = new ArrayList<>(foldingRanges.size());
+            for (FoldingRange foldingRange : foldingRanges) {
+                TextRange textRange = LSPFoldingRangeBuilder.getTextRange(
+                        foldingRange,
+                        file,
+                        document,
+                        openBraceChar,
+                        closeBraceChar
+                );
+                if ((textRange != null) && (textRange.getLength() > 1)) {
+                    int startOffset = Math.max(0, textRange.getStartOffset() - 1);
+                    int endOffset = Math.min(documentLength - 1, textRange.getEndOffset());
+                    // These ranges can tend to add whitespace at the end, so trim that before looking for braces
+                    while ((endOffset > startOffset) && Character.isWhitespace(documentChars.charAt(endOffset))) {
```

Comment:
We sometimes have to perform a pseudo-"trim" on the tail of the folding range to find the closing brace.